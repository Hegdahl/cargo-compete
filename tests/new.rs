pub mod common;

use ignore::overrides::Override;
use insta::{assert_json_snapshot, assert_snapshot};
use liquid::object;
use std::io::BufRead;

#[test]
fn atcoder_agc047() -> anyhow::Result<()> {
    let (output, tree) = run(&b""[..], "agc047")?;
    assert_snapshot!("atcoder_agc047_output", output);
    assert_json_snapshot!("atcoder_agc047_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

#[cfg(feature = "__test_with_credentials")]
#[test]
fn atcoder_practice() -> anyhow::Result<()> {
    let (output, tree) = run(common::atcoder_credentials()?, "practice")?;
    assert_snapshot!("atcoder_practice_output", output);
    assert_json_snapshot!("atcoder_practice_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

fn run(
    input: impl BufRead + 'static,
    contest: &str,
) -> anyhow::Result<(String, serde_json::Value)> {
    common::run(
        |workspace_root| -> _ {
            std::fs::create_dir_all(workspace_root.join("cargo-compete-template").join("src"))?;

            std::fs::write(
                workspace_root.join("Cargo.toml"),
                r#"[workspace]
members = ["cargo-compete-template"]
"#,
            )?;

            std::fs::write(
                workspace_root.join("compete.toml"),
                liquid::ParserBuilder::with_stdlib()
                    .build()?
                    .parse(include_str!("../resources/compete.toml.liquid"))?
                    .render(&object!({
                        "template_platform": "atcoder",
                        "submit_via_binary": false,
                    }))?,
            )?;

            std::fs::write(
                workspace_root
                    .join("cargo-compete-template")
                    .join("Cargo.toml"),
                r#"[package]
name = "cargo-compete-template"
version = "0.1.0"
edition = "2018"
"#,
            )?;

            std::fs::write(
                workspace_root
                    .join("cargo-compete-template")
                    .join("src")
                    .join("main.rs"),
                r#"fn main() {
    todo!();
}
"#,
            )?;
            Ok(())
        },
        input,
        &["", "compete", "n", contest],
        |workspace_root, output| {
            output
                .replace(workspace_root.to_str().unwrap(), "{{ cwd }}")
                .replace('/', "{{ slash_or_backslash }}")
                .replace('\\', "{{ slash_or_backslash }}")
        },
        |_| Ok(Override::empty()),
    )
}
