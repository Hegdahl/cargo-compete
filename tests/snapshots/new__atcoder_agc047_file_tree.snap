---
source: tests/new.rs
expression: tree
---
{
  "Cargo.lock": "..",
  "Cargo.toml": "[workspace]\nmembers = [\"cargo-compete-template\", \"agc047\"]\nexclude = []\n",
  "agc047": {
    "Cargo.toml": "[package]\nname = \"agc047\"\nversion = \"0.1.0\"\nedition = \"2018\"\n\n[package.metadata.cargo-compete.bin]\na = { name = \"agc047-a\", problem = { platform = \"atcoder\", contest = \"agc047\", index = \"A\", url = \"https://atcoder.jp/contests/agc047/tasks/agc047_a\" } }\nb = { name = \"agc047-b\", problem = { platform = \"atcoder\", contest = \"agc047\", index = \"B\", url = \"https://atcoder.jp/contests/agc047/tasks/agc047_b\" } }\nc = { name = \"agc047-c\", problem = { platform = \"atcoder\", contest = \"agc047\", index = \"C\", url = \"https://atcoder.jp/contests/agc047/tasks/agc047_c\" } }\nd = { name = \"agc047-d\", problem = { platform = \"atcoder\", contest = \"agc047\", index = \"D\", url = \"https://atcoder.jp/contests/agc047/tasks/agc047_d\" } }\ne = { name = \"agc047-e\", problem = { platform = \"atcoder\", contest = \"agc047\", index = \"E\", url = \"https://atcoder.jp/contests/agc047/tasks/agc047_e\" } }\nf = { name = \"agc047-f\", problem = { platform = \"atcoder\", contest = \"agc047\", index = \"F\", url = \"https://atcoder.jp/contests/agc047/tasks/agc047_f\" } }\n\n[[bin]]\nname = \"agc047-a\"\npath = \"src/bin/a.rs\"\n\n[[bin]]\nname = \"agc047-b\"\npath = \"src/bin/b.rs\"\n\n[[bin]]\nname = \"agc047-c\"\npath = \"src/bin/c.rs\"\n\n[[bin]]\nname = \"agc047-d\"\npath = \"src/bin/d.rs\"\n\n[[bin]]\nname = \"agc047-e\"\npath = \"src/bin/e.rs\"\n\n[[bin]]\nname = \"agc047-f\"\npath = \"src/bin/f.rs\"\n",
    "src": {
      "bin": {
        "a.rs": "fn main() {\n    todo!();\n}\n",
        "b.rs": "fn main() {\n    todo!();\n}\n",
        "c.rs": "fn main() {\n    todo!();\n}\n",
        "d.rs": "fn main() {\n    todo!();\n}\n",
        "e.rs": "fn main() {\n    todo!();\n}\n",
        "f.rs": "fn main() {\n    todo!();\n}\n"
      }
    }
  },
  "cargo-compete-template": {
    "Cargo.toml": "[package]\nname = \"cargo-compete-template\"\nversion = \"0.1.0\"\nedition = \"2018\"\n",
    "src": {
      "main.rs": "fn main() {\n    todo!();\n}\n"
    }
  },
  "compete.toml": "# How to manage new workspace members (\"include\" | \"exclude\" | \"focus\")\n#\n# - `skip`:    Does not modify `[workspace]`\n# - `include`: Adds the package to `workspace.members`\n# - `exclude`: Adds the package to `workspace.exclude` and create a symlink to the `compete.toml`\n# - `focus`:   Adds the package to `workspace.members` and remove the other from both of `workspace.{members, exclude}`\nnew-workspace-member = \"include\"\n\n# Path to the test file (Liquid template)\n#\n# Variables:\n#\n# - `manifest_dir`: Package directory\n# - `contest`:      Contest ID (e.g. \"abc100\")\n# - `problem`:      Problem index (e.g. \"a\", \"b\")\n#\n# Additional filters:\n#\n# - `kebabcase`: Convert to kebab case (by using the `heck` crate)\ntest-suite = \"./testcases/{{ contest }}/{{ problem | kebabcase }}.yml\"\n#test-suite = \"{{ manifest_dir }}/testcases/{{ problem | kebabcase }}.yml\"\n\n# Open files with the command (`jq` command)\n#\n# VSCode:\n#open = '[\"code\"] + (.paths | map([.src, .test_suite]) | flatten) + [\"-a\", .manifest_dir]'\n# Emacs:\n#open = '[\"emacsclient\", \"-n\"] + (.paths | map([.src, .test_suite]) | flatten)'\n\n[template]\nplatform = \"atcoder\"\nmanifest = \"./cargo-compete-template/Cargo.toml\"\nsrc = \"./cargo-compete-template/src/main.rs\"\n\n#[submit-via-binary]\n#target = \"x86_64-unknown-linux-musl\"\n##cross = \"cross\"\n#strip = \"strip\"\n##upx = \"upx\"\n",
  "testcases": {
    "agc047": {
      "a.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      5\n      7.5\n      2.4\n      17.000000001\n      17\n      16.000000000\n    out: |\n      3\n  - name: sample2\n    in: |\n      11\n      0.9\n      1\n      1\n      1.25\n      2.30000\n      5\n      70\n      0.000000001\n      9999.999999999\n      0.999999999\n      1.000000001\n    out: |\n      8\n\nextend: []\n",
      "b.yml": "---\ntype: Batch\ntimelimit: 3s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      3\n      abcxyx\n      cyx\n      abc\n    out: |\n      1\n  - name: sample2\n    in: |\n      6\n      b\n      a\n      abc\n      c\n      d\n      ab\n    out: |\n      5\n\nextend: []\n",
      "c.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      4\n      2019 0 2020 200002\n    out: |\n      474287\n  - name: sample2\n    in: |\n      5\n      1 1 2 2 100000\n    out: |\n      600013\n\nextend: []\n",
      "d.yml": "---\ntype: Batch\ntimelimit: 2s 500ms\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      3\n      2 3 1 4\n    out: |\n      121788\n  - name: sample2\n    in: |\n      2\n      1 2\n    out: |\n      36\n  - name: sample3\n    in: |\n      5\n      6 14 15 7 12 16 5 4 11 9 3 10 8 2 13 1\n    out: |\n      10199246\n\nextend: []\n",
      "e.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: >\n    out: |\n      4\n      < 0 1 8\n      + 0 1 2\n      + 2 8 2\n      + 0 0 0\n\nextend: []\n",
      "f.yml": "---\ntype: Batch\ntimelimit: 1s 250ms\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      6\n      1 8\n      6 10\n      2 7\n      4 4\n      9 3\n      5 1\n    out: |\n      5\n      0\n      7\n      5\n      0\n      0\n  - name: sample2\n    in: |\n      5\n      5 5\n      100 100\n      70 20\n      81 70\n      800 1\n    out: |\n      985\n      985\n      1065\n      1034\n      0\n  - name: sample3\n    in: |\n      10\n      2 5\n      4 4\n      13 12\n      12 13\n      14 17\n      17 19\n      22 22\n      16 18\n      19 27\n      25 26\n    out: |\n      2\n      2\n      9\n      9\n      3\n      3\n      24\n      5\n      0\n      25\n\nextend: []\n"
    }
  }
}
